### PR TITLE
feat: Col support `flex` prop

### LINKED ISSUE #7309: [Request] Additional Flex Grid props on Row/Col
### What problem does this feature solve?
I'd like to request a few additional props on the flex grid system. 
I often find the need to add some css style on the Row/Col elements (or is it the recommended way?):

### What does the proposed API look like?
On Row: 
- (flex-)direction: 'row' | 'row-reverse' | 'column' | 'column-reverse' ( ... though 'column' might confuse users due to the "Row" semantic. maybe change Row -> Grid, and Col -> Grid.Item on the next version?)
- (flex-)wrap: 'wrap' | 'no-wrap' | 'wrap-reverse'

On Col:
- (flex-)grow: number. default 0.
- align(-self): auto | start | end | center | baseline | stretch. default 'auto'

### PR DESCRIPTION
**Pull request by zombieJ**
### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 👻 What's the background?

close #7309
close https://github.com/ant-design/ant-design/issues/19728

### 💡 Solution

Add `flex` prop:
屏幕快照 2019-05-16 下午8 42 29

### 📝 Changelog

| Language | Changelog |
|---|---
| 🇺🇸 English | Col support `flex` prop |
| 🇨🇳 Chinese | Col 支持 `flex` 属性 |

### ☑️ Self Check before Merge

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

-----
[View rendered components/grid/demo/flex-stretch.md](https://github.com/ant-design/ant-design/blob/col-flex/components/grid/demo/flex-stretch.md)
[View rendered components/grid/demo/playground.md](https://github.com/ant-design/ant-design/blob/col-flex/components/grid/demo/playground.md)
[View rendered components/grid/demo/responsive-more.md](https://github.com/ant-design/ant-design/blob/col-flex/components/grid/demo/responsive-more.md)
[View rendered components/grid/demo/responsive.md](https://github.com/ant-design/ant-design/blob/col-flex/components/grid/demo/responsive.md)
[View rendered components/grid/index.en-US.md](https://github.com/ant-design/ant-design/blob/col-flex/components/grid/index.en-US.md)
[View rendered components/grid/index.zh-CN.md](https://github.com/ant-design/ant-design/blob/col-flex/components/grid/index.zh-CN.md)

### CODE REVIEW

**Comment by abenhamdine** [comment]
hi @zombieJ any news on this ?

**Comment by afc163** [comment]
https://chakra-ui.com/flex

File: components/grid/col.tsx [hunk 1]
```diff
@@ -30,6 +32,19 @@ export interface ColProps extends React.HTMLAttributes<HTMLDivElement> {
   xl?: ColSpanType | ColSize;
   xxl?: ColSpanType | ColSize;
   prefixCls?: string;
+  flex?: FlexType;
```
Comment (shaodahong) [inline components/grid/col.tsx]: 4.0 的 `Col` 和 `Row` 是不是默认 `flex` 比较好，最低 IE 11 浏览器，flex 使用没问题吧

Comment (zombieJ) [reply components/grid/col.tsx]: 也不是不可以~

Comment (shaodahong) [reply components/grid/col.tsx]: 那我先建个 Issue 说下改造？等空闲了修改下

**Comment by akaguny** [comment]
any news?

File: components/grid/style/mixin.less [hunk 2]
```diff
@@ -29,27 +29,6 @@
...
-  .col(@index, @list) when (@index > @grid-columns) {
-    // terminal
-    @{list} {
-      flex: 0 0 auto;
-      float: left;
-    }
-  }
-  .col(1); // kickstart it
-}
-
```
Comment (afc163) [inline components/grid/style/mixin.less]: Fanally, I believe this can help to reduce at least 50Kb css size.

**Comment by shaodahong** [comment]
- [x] remove **Row** `type` prop, now default by `flex` mode
- [x] remove .less file float style

**Comment by shaodahong** [comment]
@afc163 @zombieJ this request https://github.com/ant-design/ant-design/issues/7309#issuecomment-546609259 is ok?

File: components/grid/style/mixin.less [hunk 3]
```diff
@@ -29,27 +29,6 @@
...
 .loop-grid-columns(@index, @class) when (@index > 0) {
```
Comment (afc163) [inline components/grid/style/mixin.less]: Can we remove all generated loop code？

Comment (shaodahong) [reply components/grid/style/mixin.less]: I will try it later

Comment (shaodahong) [reply components/grid/style/mixin.less]: `Grid` base on 24 range, so `.loop-grid-column` class is important, but I change it with `flex`, get a better experience

**Comment by zombieJ** [comment]
https://deploy-preview-16635--ant-design.netlify.com/components/grid-cn/#components-grid-demo-flex

`sub-element monospaced arrangement` not correct

**Comment by zombieJ** [comment]
两行 Row 之间的间距变化是因为什么？
另外，原本的灰色背景可以加一下，对齐比较好看些

**Comment by shaodahong** [comment]
@zombieJ 

> Why padding changed between the rows?

look like 3.x `flex` mode use `gutter` bug

https://codesandbox.io/s/suspicious-swirles-lw34r

> And the gray background is better to add back.

yep, 3.x version we define some `ant-row-flex` style in the site, now we remove `-flex`, and I back to this now

**Comment by zombieJ** [comment]
Form line mode not correct. Please check:

Current:
https://next.ant.design/components/form-cn/#components-form-demo-layout

With this PR:
https://deploy-preview-16635--ant-design.netlify.com/components/form-cn/#components-form-demo-layout

**Comment by shaodahong** [comment]
@zombieJ plz double check, and prepare merge

**Comment by zombieJ** [comment]
标题 `Flex 布局` 可以去掉，作为 `概述` 的额外描述。demo 里的 flex 文字也可以去掉了。

**Comment by afc163** [comment]
We can hint ``type="flex"` is not neccessary now blahblah` if set.

**Comment by zombieJ** [comment]
> We can hint ``type="flex"` is not neccessary now blahblah` if set.

Can move into `Migrate to v4` like Table: https://next.ant.design/components/table/#Migrate-to-v4

**Comment by shaodahong** [comment]
> > We can hint ``type="flex"` is not neccessary now blahblah` if set.
> 
> Can move into `Migrate to v4` like Table: https://next.ant.design/components/table/#Migrate-to-v4

also, and we 4.0 release doc can metion this too

**Comment by shaodahong** [comment]
Wait for https://github.com/ant-design/ant-design/pull/16635#issuecomment-551093001 this list all checked.

**Comment by shaodahong** [comment]
> * [x]  Remove **Row** `type` prop, now default by `flex` mode
> * [x]  Remove .less file float style
> * [x]  Fix flex with `gutter` invalid vertical
> * [x]  Doc (we can reorganize Grid doc )
>   
>   * [x]  Remove FAQ
>   * [x]  Remove layout and demo description metion with flex(now default `flex`)

All of my changed list.

Done.

File: components/grid/index.en-US.md [hunk 4]
```diff
@@ -90,17 +88,17 @@ If the Ant Design grid layout component does not meet your needs, you can use th
...
-| type | layout mode, optional `flex`, [browser support](http://caniuse.com/#search=flex) | string |  |  |
+| align | vertical alignment | `top` `middle` `bottom` | `top` |  |
+| gutter | spacing between grids, could be a number or a object like `{ xs: 8, sm: 16, md: 24}`. or you can use array to make horizontal and vertical spacing work at the same time `[horizontal, vertical]` (supported after `3.24.0`) | number/object/array | 0 |  |  |
+| justify | horizontal arrangement | `start` `end` `center` `space-around` `space-between` | `start` |  |
 
 ### Col
 
 | Property | Description | Type | Default | Version |
 | --- | --- | --- | --- | --- |
+| flex | flex layout style | string \| number | - | 3.19.0 |
```
Comment (zombieJ) [inline components/grid/index.en-US.md]: 这个不需要版本号了

**Comment by zombieJ** [comment]
LGTM

File: components/grid/index.en-US.md (hunk 4 above)
Comment (shaodahong) [reply components/grid/index.en-US.md]: Done.

**Review by afc163 — APPROVED** [review]

File: components/grid/style/index.less [hunk 5]
```diff
@@ -21,52 +14,53 @@
...
 .@{ant-prefix}-col {
   position: relative;
+  flex: 1 auto;
+  max-width: 100%;
   // Prevent columns from collapsing when empty
   min-height: 1px;
 }
 
-.make-grid-columns();
 .make-grid();
```
Comment (afc163) [inline components/grid/style/index.less]: 感觉这个也能干掉？

Comment (shaodahong) [reply components/grid/style/index.less]: 干不掉，之前干掉然后又捡回来了，这个就是栅格的基础，没有这个就没有栅格了……

```css
display: block;
flex: 0 0 percentage((@index / @grid-columns));
max-width: percentage((@index / @grid-columns));
```

**Pull request merged**